Simplify and optimize worker task scheduling

mourner opened the pull request:
Closes #10187. There are two independent commits here. 

The first one simplifies `setData` coalescing logic in GeoJSON source, previously introduced in #5902. I did this first to make worker logic easier to reason about, but theoretically it should improve performance — by moving the coalescing logic to the main thread, we avoid flooding the worker message queue with tasks that will get discarded. Technically the flow and consequently the overall performance characteristics shouldn't change, as I tried to demonstrate on this very rough and clunky chart:

![Image from iOS (1)](https://user-images.githubusercontent.com/25395/109312849-32cdd880-7850-11eb-8f34-b73a75eb093c.jpg)

Previously, any `setData` calls that follow one that's still in progress would get sent to the worker, which remembers the last call while returning the previous one as "abandoned", and waits until that first `setData` call successfully finishes processing and then calls `coalesce` message that tells the worker to additionally do an update for the last `setData` call it caught before.

Now, we simply don't send any worker messages on additional `setData` calls while a `setData` is already in progress, but we remember to issue one more `setData` with the last updated `data` if there are any updates which we previously refused. 

The second commit is the one that fixes the Safari performance issue — it was caused by Safari being too slow to process worker tasks that are delayed to run after the current event loop (introduced in #8633), which we do to make sure `<cancel>` messages are processed _before_ the tasks that were cancelled if both come together to the worker in a batch of `postMessage` calls. Previously, all messages were delayed, then #8913 made it delay only on the worker side, and finally #9031 introduced an explicit `actor.send` parameter to additionally delay `getResource` on the main thread to fix a perf regression. This PR changes the logic so that messages are handled immediately _by default_, and only delayed for `<cancel>` processing explicitly for calls where we commonly expect cancellations (mostly network-related) — in this case `loadTile`, `loadDEMTile` and `getResource`. I confirmed that this fixes the Safari `setData` performance issue in the ticket, while not increasing the percentage of unfulfilled cancellations when browsing the map quickly (it's about 40% before and after the PR).

Tagging @kkaefer @ChrisLoer just in case because it affects the code you added significantly — take a look if you have time but no worries if not.

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - write tests for all new functionality (seems sufficiently covered by existing tests)
 - [x] post benchmark scores
 - [x] manually test the debug page
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Fixes a performance regression in Safari on frequent GeoJSON setData calls</changelog>`

mourner commented:
Internal benchmarks show no difference in performance, which is expected:

![image](https://user-images.githubusercontent.com/25395/109327946-8e549200-7861-11eb-9c82-7117adff12a8.png)

ansis commented on src/source/geojson_source.js:
```diff
@@ -276,24 +250,28 @@ class GeoJSONSource extends Evented implements Source {
+
+            if (err) {
+                this.fire(new ErrorEvent(err));
                 return;
```
An error here will prevent the `_updateWorkerData()` call when `_coalesce = true`. Does this matter?

mourner replied:
Nice catch! It's a rare occasion (e.g. running `setData` with a malformed GeoJSON and then following it shortly with a correct one), but fixed it just in case.

ansis reviewed (commented):
The geojson simplification is great.

It looks like the callbacks from [`getGlyphs`](https://github.com/mapbox/mapbox-gl-js/blob/main/src/source/worker_tile.js#L166) and `getImages` aren't getting added to the scheduler when I think they should be. This prioritization would matter if the glyphs for one time loaded before the vectortile of another tile. Our benchmarks don't appear to be covering this case. But maybe that means it's fine as is.

By bypassing the scheduler a whole bunch of work is now not measured as part of the `workerTask` diagnostic metric. This could be fixed by adding something like [this](https://github.com/mapbox/mapbox-gl-js/blob/c4229dce5d303b2abfdf926185d8530682904dea/src/util/scheduler.js#L36) to the other path or by putting all work through the scheduler (and making some of it immediate there).

The scheduler already gave messages the [highest priority](https://github.com/mapbox/mapbox-gl-js/blob/c4229dce5d303b2abfdf926185d8530682904dea/src/util/scheduler.js#L89) by default. Making all those immediate would probably work well.

pepe-invest-git reviewed (approved):
\

mourner commented:
> It looks like the callbacks from [`getGlyphs`](https://github.com/mapbox/mapbox-gl-js/blob/main/src/source/worker_tile.js#L166) and `getImages` aren't getting added to the scheduler when I think they should be. This prioritization would matter if the glyphs for one time loaded before the vectortile of another tile. Our benchmarks don't appear to be covering this case. But maybe that means it's fine as is.

Good point! I also didn't realize responses after returning the result from the other thread also get scheduled, so I restored the condition that this only happens on the worker side, and additionally added `mustQueue` flags for the `getGlyphs` and `getImages` calls to make sure the behavior doesn't change.

> By bypassing the scheduler a whole bunch of work is now not measured as part of the `workerTask` diagnostic metric. This could be fixed by adding something like [this](https://github.com/mapbox/mapbox-gl-js/blob/c4229dce5d303b2abfdf926185d8530682904dea/src/util/scheduler.js#L36) to the other path or by putting all work through the scheduler (and making some of it immediate there).

👍  went with the first option.

mourner commented:
For the record, benchmark results after addressing feedback:

![image](https://user-images.githubusercontent.com/25395/110344307-070fd700-8036-11eb-9242-7170a605a9ce.png)

ansis commented on src/source/worker_tile.js:
```diff
@@ -169,7 +169,7 @@ class WorkerTile {
                     maybePrepare.call(this);
                 }
-            }, undefined, undefined, taskMetadata);
+            }, undefined, true, taskMetadata);
```
I don't think we need to queue this on the main thread. We need to queue the response to this

mourner replied:
So, the intent is to make `mustQueue` force queing only on the worker thread, whether it's the task (if you're sending from the main) or the response (if you're sending from the worker).

ansis commented on src/util/actor.js:
```diff
@@ -107,20 +108,22 @@ class Actor {
-                // executing the next task in our queue, postMessage preempts this and <cancel>
-                // messages can be processed. We're using a MessageChannel object to get throttle the
-                // process() flow to one at a time.
+            if (isWorker() && data.mustQueue) {
```
We need to pass the responses from getImages and getGlyphs to the scheduler. Checking for the presence of `callback.metadata` in actor.js might be enough to decide whether to do that but I think letting the scheduler decide that might be slightly cleaner

The queuing on the main thread was [intentional](https://github.com/mapbox/mapbox-gl-js/pull/9031) but it looks like it might not be needed since we dropped IE. I think [this]((https://github.com/mapbox/mapbox-gl-js/blob/719301fcae7f359453cdfc0bf699a40734270cae/src/util/ajax.js#L245-L246)) is the only case where we did queuing on the main thread. It was also applied to iOS Safari < 12.1 but I don't think it was actually  needed there... not sure though.  @arindam1993 do you remember if the queuing was only needed for IE?

arindam1993 replied:
Also old Safari verions wherein `AbortController` doesn't actually abort fetches.

ansis commented:
> I also didn't realize responses after returning the result from the other thread also get scheduled, so I restored the condition that this only happens on the worker side, and additionally added mustQueue flags for the getGlyphs and getImages calls to make sure the behavior doesn't change.

From what I can tell it doesn't actually apply queuing to the responses to these calls. The messages have `type === "<response>"`. I left notes in the two places I think we need changes

mourner commented:
@ansis OK, so the `mustQueue` (that I intended to only affect the response on the wokrer) wasn't passed in `getGlyphs` and `getImages` because of an existing blunder in the code (see commit earlier), but for now I just reverted making the responses queueing because the logic is getting too confusing now. Maybe we should merge it without glyph/image queueing for now, but plan a more substantial refactor as a follow-up — e.g. use options instead of a long list of arguments to `send`, and maybe separately have `mustQueue` and `mustQueueCallback` options.

arindam1993 commented:
> @ansis OK, so the `mustQueue` (that I intended to only affect the response on the wokrer) wasn't passed in `getGlyphs` and `getImages` because of an existing blunder in the code (see commit earlier), but for now I just reverted making the responses queueing because the logic is getting too confusing now. Maybe we should merge it without glyph/image queueing for now, but plan a more substantial refactor as a follow-up — e.g. use options instead of a long list of arguments to `send`, and maybe separately have `mustQueue` and `mustQueueCallback` options.

This! Do you think its worth moving all the queueing, throttling and cancellation logic into `ajax.js\RequestManager` layer to simplify some of this down.
Do we need that kind of functionality for anything other than network requests?
